### PR TITLE
int audit core::{hash, fmt}, std::sync

### PR DESCRIPTION
**Pull request by nrc**
fmt and hash are pretty straightforward I think. sync is a bit more complex. I thought one or two of the `isize`s ought to be `i32`s, but that would require a bunch of casting (the root cause being the lack of atomics other than isize/usize).

r? @alexcrichton

### CODE REVIEW

**Comment by pnkfelix** [comment]
cc #22240

**Comment by alexcrichton** [comment]
> I thought one or two of the isizes ought to be i32s, but that would require a bunch of casting (the root cause being the lack of atomics other than isize/usize).

I think these are fine yeah, they're not showing up in the public API and we could perhaps add an `AtomicU32` later in the future.

**Comment by Manishearth** [comment]
Needs https://github.com/Manishearth/rust/commit/a5009b8f1498ae4b7fabb129fb211ee5bb35a0c2 to compile on master.

**Comment by Manishearth** [comment]
Also needs https://github.com/rust-lang/rust/commit/b182cd7245a999ac702f88c89dcc28811d6fdf8a

Might want to take a look at that (it only touches tests), since your commit message mentions some trickiness in sync.

File: src/libcore/fmt/mod.rs [hunk 1]
```diff
@@ -731,9 +746,9 @@ impl Display for char {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Pointer for *const T {
     fn fmt(&self, f: &mut Formatter) -> Result {
-        f.flags |= 1 << (FlagV1::Alternate as uint);
-        let ret = LowerHex::fmt(&(*self as uint), f);
-        f.flags &= !(1 << (FlagV1::Alternate as uint));
+        f.flags |= 1 << (FlagV1::Alternate as u32);
+        let ret = LowerHex::fmt(&(*self as u32), f);
```
Comment (SSheldon) [inline src/libcore/fmt/mod.rs]: Casting the pointer to `u32` loses part of it on 64bit, see #22854. Should this be usize instead?

**Pull request merged**